### ISSUE TITLE
stub exited while waiting for connection: exit status 0

### ISSUE THREAD

**Issue by ameetchhatwal** [go-delve/delve#1707]
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)?

```
Delve Debugger
Version: 1.3.0
Build: $Id: 2f59bfc686d60989dcef9de40b480d0a34aa2fa5 $
```

2. What version of Go are you using? (`go version`)?
`go version go1.13.1 darwin/amd64```
`

3. What operating system and processor architecture are you using?
mac osX 10.14.6

4. What did you do?
When I attached a process in GoLand the first time it worked fine and after some time its started failing with error below. 
Then I used the command line dlv command dlv attach <pid> and its giving the same error. 

5. What did you expect to see?

6. What did you see instead?

GOLAND error 
```/Applications/GoLand.app/Contents/plugins/go/lib/dlv/mac/dlv --listen=localhost:50393 --headless=true --api-version=2 attach 28780 -- #gosetup
API server listening at: 127.0.0.1:50393
could not attach to pid 28780: stub exited while waiting for connection: exit status 0

Debugger finished with exit code 1```

CMD line error 
```could not attach to pid 28780: stub exited while waiting for connection: exit status 0```



I have checked my hosts file and it looks fine. I am not sure what went wrong since its worked the first time and not it doesnt. 

Please help

**Comment by aarzilli** [comment]
You need to try running debug server directly:

```
/Library/Developer/CommandLineTools/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/debugserver --attach=pid
```

and see what happens.

**Comment by ameetchhatwal** [comment]
> You need to try running debug server directly:
> 
> ```
> /Library/Developer/CommandLineTools/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/debugserver --attach=pid
> ```
> 
> and see what happens.

**Comment by aarzilli** [comment]
Try: `/Library/Developer/CommandLineTools/Library/PrivateFrameworks/LLDB.framework/Versions/A/Resources/debugserver 127.0.0.1:9000 --attach=<pid>`

**Comment by ameetchhatwal** [comment]
I get this 
`debugserver-@(#)PROGRAM:debugserver  PROJECT:debugserver-1001.0.13.3
 for x86_64.
Attaching to process 10608...
error: failed to attach process 10608: unable to start the exception thread
Exiting.`

**Comment by aarzilli** [comment]
It's a problem with debugserver, you should report it to apple.

**Comment by ameetchhatwal** [comment]
Thanks a lot. But When I initially tried attaching the process from GoLand it worked the first time and stopped running afterwards. Could there be anything else which might be stopping `dlv` cmd to run from GoLand and Terminal ?

**Comment by aarzilli** [comment]
What's stopping dlv from working is debugserver not working.

**Comment by ameetchhatwal** [comment]
Gotcha! Not sure why I ran initially and what caused it to stop working. Thanks a lot for for explaining it!

**Comment by aarzilli** [comment]
Close, upstream issue.

**Comment by matthewhembree** [comment]
For whomever stumbles upon this thread:

This is due to a MacOS restriction.
`taskgated` (Gatekeeper) checks for an entitlement `com.apple.security.get-task-allow` to attach the debugger. You can observe the relevant logs in Console.app for the processes (taskgated, debugserver, and kernel).

I was able to resolve this issue for me by running `dlv attach...` with `sudo`.

Or if you cannot run `dlv` as root (i.e. you really want to clicky attach in GoLand):
1. Create a code signing certificate with [Keychain Access](https://support.apple.com/guide/keychain-access/create-self-signed-certificates-kyca8916/mac)
2. Create an `entitlements.plist` file containing:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
         <key>com.apple.security.get-task-allow</key>
         <true/>
</dict>
</plist>
```
3. Notarize your executable (must be done after every build):
```sh
codesign -fs <certificate name> --timestamp --options=runtime --entitlements entitlements.plist $GOPATH/bin/<executable>
```
4. Attach `dlv` to the process.

TL;DR: Run `dlv attach` with `sudo`.

P.S
Like the OP, I was also able to attach Delve on the initial attempt. It was probably allowed due to a bug. I was able to do it again one time after disabling SIP(`csrutil disable`).

P.P.S. There's probably another way to get this running easily in GoLand. Maybe renaming `/Applications/GoLand.app/Contents/plugins/go/lib/dlv/mac/dlv` and replacing it with a script named `dlv` that calls the former `dlv` with `sudo` and passes the args. OR, maybe I'll open an issue with JetBrains.

**Comment by ravron** [comment]
I ran into the exact issue described by @matthewhembree, but was perplexed that coworkers with a similar configuration had the same problem. I don't fully understand the root of the issue, but seems like the `debugserver` installed by Apple's standalone CommandLineTools package isn't properly signed in some way. I found that all the people for whom `debugserver` had stopped working shared the following feature:

```
$ xcode-select -p
/Library/Developer/CommandLineTools
```

whereas those for whom it kept working had

```
$ xcode-select -p
/Applications/Xcode.app/Contents/Developer
```

So I did the following:

1. [Uninstall CommandLineTools](https://developer.apple.com/library/archive/technotes/tn2339/_index.html#//apple_ref/doc/uid/DTS40014588-CH1-HOW_CAN_I_UNINSTALL_THE_COMMAND_LINE_TOOLS_): `sudo rm -rf /Library/Developer/CommandLineTools`
2. Install Xcode from the App Store
3. Run `xcode-select --install`
4. Open Xcode and accept the prompts to install additional tools

And debugging in GoLand works again!

**Comment by shanemcd** [comment]
I just ran into this and fixed by running:

```
$ sudo /usr/sbin/DevToolsSecurity --enable
```

**Comment by hilaily** [comment]
I finally get this work.

1. I tried to enable the DevToolsSecurity , but it does not work.
```bash
sudo /usr/sbin/DevToolsSecurity --enable
```

2. Then I run xcode-select-p and got this.
```
xcode-select -p
/Library/Developer/CommandLineTools
```
After installing the Xcode, CommandLineTools is in the other direction, but it still does not work.
 
3. Then I found the idea by @matthewhembree
> P.P.S. There's probably another way to get this running easily in GoLand. Maybe renaming /Applications/GoLand.app/Contents/plugins/go/lib/dlv/mac/dlv and replacing it with a script named dlv that calls the former dlv with sudo and passes the args. OR, maybe I'll open an issue with JetBrains.

Create a shell script like this
```
#!/bin/bash
echo my-password | sudo -S -E my-dlv-command ${@}
```
Make a sort link for it and replace dlv command. (my Goland use my own dlv command instead of that in the Goland directory).
and then everything works well.

**Comment by nd** [comment]
@hilaily it should work without sudo. Please check if running the following command helps:

`sudo dscl . append /Groups/_developer GroupMembership <your username>`

**Comment by hilaily** [comment]
@nd cool, it works well when I use the origin dlv command, thanks :)

**Comment by AndrewRPorter** [comment]
@matthewhembree what PID am I attaching delve to? I am running into this issue in VSCode.

**Comment by matthewhembree** [comment]
> @matthewhembree what PID am I attaching delve to? I am running into this issue in VSCode.

@AndrewRPorter The PID will be of your application that you want to debug.

OT: I've found it difficult to do serious debugging with VSCode. The inspector couldn't handle large structs. It wouldn't show the values hierarchically and it would truncate large structs. If VSCode meets your needs, then stick with it. VSCode is less demanding on resources, so that's a plus.

**Comment by jachands** [comment]
For anyone else that may run into this: If you're on an M1 Mac, make sure you have the arm64 version of VS Code installed - that fixed it for me.

**Comment by jahands** [comment]
For anyone else that may run into this: If you're on an M1 Mac, make sure you have the arm64 version of VS Code installed - that fixed it for me.

**Comment by charconstpointer** [comment]
> I just ran into this and fixed by running:
> 
> ```
> $ sudo /usr/sbin/DevToolsSecurity --enable
> ```

Fixed my issues with remote dev setup on a M1 Pro 🥳

**Comment by guiferpa** [comment]
🐛 

Unfortunately I still on the same issue. Somebody already tried `dlv` from terminal, like this pic below?

<img width="1440" alt="Screen Shot 2022-09-02 at 15 17 21" src="https://user-images.githubusercontent.com/9096630/188214141-2a26cf2a-d7e2-45d6-9f40-10824713183e.png">

> ⚠️  I'm using a MacOs with M1 Chip 😢 

### `go env`

```
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/OlaIsaac/Library/Caches/go-build"
GOENV="/Users/OlaIsaac/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/OlaIsaac/.asdf/installs/golang/1.17.6/packages/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/OlaIsaac/.asdf/installs/golang/1.17.6/packages"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/OlaIsaac/.asdf/installs/golang/1.17.6/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/OlaIsaac/.asdf/installs/golang/1.17.6/go/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.17.6"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/OlaIsaac/Projects/src/github.com/OlaIsaac/backoffice-api-go/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/4q/z0nsw4x13rj2zbw55wy8v37w0000gp/T/go-build4141306592=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
